standard/expressions.md

```diff
@@ -2227,17 +2228,99 @@ If the *primary_no_array_creation_expression* of an *element_access* is a value
+  - The result of the transformation becomes the result of the array access.
+
+> *Example*: Given the following one-dimensional array and Index:
+>
+> <!-- Example: {template:"code-in-main", name:"ArrayAccessWithIndex1", expectedOutput:["green"], expectedException:"IndexOutOfRangeException"} -->
+> ```csharp
+> string[] words = new string[] { "red", "green", "blue" };
+> Index idx = 1;
+> Console.WriteLine(words[idx]);      // green
+> Console.WriteLine(words[^0]);       // IndexOutOfRangeException
+> ```
+>
+> `words[idx]` is transformed by the implementation to `words[idx.GetOffset(words.Length)]`. Similarly, `words[^0]` is transformed by the implementation to `words[(^0).GetOffset(words.Length)]`, which results in an exception.
+>
+> Given the following jagged array and Indexes:
+>
+> <!-- Example: {template:"code-in-main", name:"ArrayAccessWithIndex2", expectedOutput:["17"]} -->
+> ```csharp
+> int[][] values = new int[][] { new int[] { 10, 9 }, new int[] { 6, 12, 17 }};
+> Index idx1 = 1;
+> Index idx2 = ^1;
+> Console.WriteLine(values[idx1][idx2]);      // 17 (values[1][2])
+> ```
+>
+> `values[idx1][idx2]` is transformed by the implementation to
+>
+> ```csharp
+> values[idx1.GetOffset(values.Length)][idx2.GetOffset(values[idx1].Length)]
+> ```
+>
+> Given the following multidimensional array and Indexes:
+>
+> <!-- Example: {template:"code-in-main", name:"ArrayAccessWithIndex3", expectedErrors:["CS0029","CS0029"]} -->
+> ```csharp
+> int[,] values2D = {{10, 5, 7, 1}, {34, 13, 6, 2}};
+> Index idx3 = 1; 
+> Index idx4 = ^1;
+> Console.WriteLine(values2D[idx3, idx4]);      // won't compile!
+> ```
+>
+> as there is no implicit conversion from `System.Index` to `int`, what one might like to express simply as `values2D[idx3, idx4]` must instead be written explicitly as
+>
+> ```csharp
+> values2D[idx3.GetOffset(values2D.GetUpperBound(0) + 1), idx4.GetOffset(values2D.GetUpperBound(1) + 1)]
+>
+> ```
+>
+> Given the following one-dimensional array:
+>
+> <!-- Example: {template:"code-in-main", name:"ArrayAccessWithIndex4"} -->
+> ```csharp
+> string[] seasons = new string[] { "Summer", "Autumn", "Winter", "Spring" };
+> string[] names = seasons[0..2]; // slice containing "Summer" and "Autumn"
+> ```
+>
+> `seasons[0..2]` is transformed by the implementation to
+>
+> <!-- NotAn$Example: {} -->
+> ```csharp
+>  System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray<string>(seasons, 0..2)
+> ```
+>
+> which returns a `string[]` slice.
+>
+> Given the following jagged array:
+>
+> <!-- Example: {template:"code-in-main", name:"ArrayAccessWithIndex5", expectedOutput:["42"]} -->
+> ```csharp
+> int[][] values = new int[][] { new int[] { 10, 9, 5 },
+>    new int[] { 6, 12, 17, 32 }, new int[] { 28, 42 } };
+> Console.WriteLine(values[1..3][^1][..2][^1]);      // 42
+> ```
+>
+> all the Range and Index expressions in `values[1..3][^1][..2][^1]` are transformed by the implementation, resulting in `values[2][1]`, which is 42. *end example*
```

Comment:
Because this is an example, let's consider writing out each transformation:  `values[1..3]` evaluates to .... Then `result`[^1]` evaluates to `result2` which is .... Then, `result2[..2]` evaluates to `result3` which is .... Finally, `result3[^1]` is 42.